### PR TITLE
structureclimbing turned component.

### PR DESCRIPTION
**Pull request by nemvar**
## About The Pull Request

Turns the structure climbing functionality of stuff like tables into a component.

## Why It's Good For The Game

The potential benefits of this is that it would be possible to apply this behaviour to stuff like machines or effects.

The mousedropping of held items to drop them onto tables is now handled by an element. Please tell me if I NEED to convert everything related to table stuff.

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Nice. This will be useful for rails when we get around to multi-z

**Comment by nemvar** [comment]
You could even apply this to turfs or something if you really wanted to.

**Comment by cacogen** [comment]
this shit is cool

File: code/datums/components/climbable.dm [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+	var/adjusted_climb_time = climb_time
+	if(user.restrained()) //climbing takes twice as long when restrained.
+		adjusted_climb_time *= 2
+	if(isalien(user))
+		adjusted_climb_time *= 0.25 //aliens are terrifyingly fast
+	if(HAS_TRAIT(user, TRAIT_FREERUNNING)) //do you have any idea how fast I am???
+		adjusted_climb_time *= 0.8
+
+	structureclimbers |= user
```
Comment (ninjanomnom) [inline code/datums/components/climbable.dm]: do a keyed list and you can also skip all this if they're already climbing

File: code/datums/components/climbable.dm [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	RegisterSignal(parent, list(COMSIG_ATOM_FORCED_MOVEMENT_BUMPED), .proc/on_forced_movement_bump)
+	RegisterSignal(parent, list(COMSIG_MOUSEDROPPED_ONTO), .proc/on_MouseDrop_T)
+	RegisterSignal(parent, list(COMSIG_ATOM_ATTACK_HAND), .proc/on_attack_hand)
+	if(istype(parent, /obj/structure/closet))
+		RegisterSignal(parent, list(COMSIG_CLOSET_CLOSE), .proc/on_closet_close)
+		RegisterSignal(parent, list(COMSIG_CLOSET_OPEN), .proc/on_closet_open)
+
+///Handles the actual act of moving
+/datum/component/climbable/proc/do_climb(atom/source, atom/movable/A)
+	A.forceMove(get_step_towards(A, source))
```
Comment (ninjanomnom) [inline code/datums/components/climbable.dm]: You need to check somewhere that there isn't something else preventing the movement

File: code/datums/forced_movement.dm [hunk 3]
```diff
@@ -87,7 +87,5 @@
 
 /mob/Bump(atom/A)
 	. = ..()
-	if(force_moving && force_moving.allow_climbing && isstructure(A))
-		var/obj/structure/S = A
-		if(S.climbable)
-			S.do_climb(src)
+	if(force_moving) //this should be turned into a component.
+		SEND_SIGNAL(A, COMSIG_ATOM_FORCED_MOVEMENT_BUMPED, src, force_moving)
```
Comment (ninjanomnom) [inline code/datums/forced_movement.dm]: Listen to the bump signal, dont make a new one here

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/climbable.dm (hunk 2 above)
Comment (nemvar) [reply code/datums/components/climbable.dm]: Reimplemented the old hacky thing because I couldn't think of anything better

**Comment by Jalleo** [comment]
Crates are climb able too right?

**Comment by nemvar** [comment]
Yas

File: code/datums/components/climbable.dm [hunk 4]
```diff
@@ -0,0 +1,110 @@
+///This components makes it possible to climb (forcemove yourself) onto an atom by clicking dragging onto it.
+/datum/component/climbable
+	///How long the climb act takes.
+	var/climb_time
+	///How long the climb act stuns the climber
+	var/climb_stun
+	///If the parent exhibits tablelike behaviour. Basically this means that by dragging your active item onto it, you place it onto the table.
+	var/table_behaviour
```
Comment (AnturK) [inline code/datums/components/climbable.dm]: That feels like completely unrelated functionality.

**Comment by nemvar** [comment]
If someone is angry that I didn't convert all the table behaviour into the element, tell me while I'm still willing to work on it.

File: code/datums/elements/table_behaviour.dm [hunk 5]
```diff
@@ -0,0 +1,19 @@
+///Very simplistic element. Basically only makes it so that if you mousedrop a held item onto parent, it gets moved to that location.
+///Eventually, one could outsource all table behaviour to this. Be it tableing or CanPass().
+/datum/element/tablebehaviour
+
+/datum/element/tablebehaviour/Attach(datum/target)
```
Comment (AnturK) [inline code/datums/elements/table_behaviour.dm]: Shouldn't this also do the basic click with item to place ?
